cmd/influx_inspect/inmem2tsi/inmem2tsi.go

```diff
@@ -103,61 +97,147 @@ func (cmd *Command) run(path, walPath string, verbose bool) error {
-
-		mm, err := inmemIndex.Measurement(name)
-		if err != nil {
+	// Write out tsm1 files.
+	cmd.Logger.Info("iterating over tsm files")
+	for _, path := range tsmPaths {
+		cmd.Logger.Info("processing tsm file", zap.String("path", path))
+		if err := cmd.processTSMFile(tsiIndex, path); err != nil {
 			return err
-		} else if mm == nil {
-			return nil
 		}
+	}
 
-		if err := mm.ForEachSeriesByExpr(nil, func(tags models.Tags) error {
-			cmd.Logger.Info("series", zap.String("name", string(name)), zap.String("tags", tags.String()))
-			if err := tsiIndex.CreateSeriesIfNotExists(nil, name, tags); err != nil {
-				return fmt.Errorf("cannot create series: %s %s (%s)", name, tags.String(), err)
-			}
-			return nil
-		}); err != nil {
+	// Write out wal files.
+	cmd.Logger.Info("iterating over wal files")
+	for _, path := range walPaths {
+		cmd.Logger.Info("processing wal file", zap.String("path", path))
+		if err := cmd.processWALFile(tsiIndex, path); err != nil {
 			return err
 		}
-		return nil
-
-	}); err != nil {
-		return err
 	}
 
-	cmd.Logger.Info("compacting index")
-
 	// Attempt to compact the index & wait for all compactions to complete.
+	cmd.Logger.Info("compacting index")
 	tsiIndex.Compact()
 	tsiIndex.Wait()
 
-	cmd.Logger.Info("closing tsi index")
-
 	// Close TSI index.
+	cmd.Logger.Info("closing tsi index")
 	if err := tsiIndex.Close(); err != nil {
 		return err
 	}
 
-	cmd.Logger.Info("moving tsi to permanent location")
-
 	// Rename TSI to standard path.
+	cmd.Logger.Info("moving tsi to permanent location")
 	if err := os.Rename(tmpPath, indexPath); err != nil {
 		return err
 	}
 
 	return nil
 }
 
+func (cmd *Command) processTSMFile(index *tsi1.Index, path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	r, err := tsm1.NewTSMReader(f)
+	if err != nil {
+		cmd.Logger.Warn("unable to read, skipping", zap.String("path", path), zap.Error(err))
+		return nil
+	}
+	defer r.Close()
+
+	for i := 0; i < r.KeyCount(); i++ {
+		key, _ := r.KeyAt(i)
+		seriesKey, _ := tsm1.SeriesAndFieldFromCompositeKey(key)
+		name, tags := models.ParseKey(seriesKey)
+
+		if cmd.Verbose {
+			cmd.Logger.Info("series", zap.String("name", string(name)), zap.String("tags", tags.String()))
+		}
+
+		if err := index.CreateSeriesIfNotExists(nil, []byte(name), tags); err != nil {
+			return fmt.Errorf("cannot create series: %s %s (%s)", name, tags.String(), err)
+		}
+	}
+	return nil
+}
+
+func (cmd *Command) processWALFile(index *tsi1.Index, path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	r := tsm1.NewWALSegmentReader(f)
+	defer r.Close()
+
+	for r.Next() {
+		entry, err := r.Read()
+		if err != nil {
+			n := r.Count()
+			cmd.Logger.Warn("wal file corruption", zap.String("path", path), zap.Int64("n", n))
+			break
+		}
+
+		switch t := entry.(type) {
+		case *tsm1.WriteWALEntry:
```

Comment:
I think this will create an inconsistent index because it will index everything in the WAL when there could be deletes stored later.  Maybe instead, use the `CacheLoader` to re-create a `Cache` and iterate over the keys that still remain in the cache?

Reply:
I switched it to use the `tsm1.CacheLoader` instead in 22a0e42. Can you take a look?